### ISSUE TITLE
[Proposal] Implicit Default Optional Parameter 

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/vblang#29]
[(Ported from Roslyn Repo)](https://github.com/dotnet/roslyn/issues/10293)

The grammar definition for an `Optional Parameter` is something similar to this.

```
OptionalParameter ::= "Optional" ParameterName Typing? DefaultToValue?
           Typing ::= "As" TypeIdentifier
   DefaultToValue ::= "=" ( "Nothing" | ConstantValue )
```

**Examples**

``` vb.net
Foo( Optional arg0 ) ' --> Foo( Optional arg0 As Object = Nothing )
Foo( Optional arg1 As String ) ' --> Foo( Optional arg1 As String = Nothing )
Foo( Optional arg2 As String = Nothing ) ' --> Foo( Optional arg2 As String = Nothing )
Foo( Optional arg3 = "" ) ' --> Foo( Optional arg3 As String = "" )
Foo( Optional arg4 As Integer ) ' --> Foo( Optional arg4 As Integer = Nothing )
Foo( Optional arg5 As Integer = 0 ) ' --> Foo( Optional arg5 As Integer = 0 )
Foo( Optional arg6 As Integer = 1 ) ' --> Foo( Optional arg6 As Integer = 1 )
Foo( Optional arg7 = 7 ) ' --> Foo( Optional arg7 As Integer = 7 )
```

It will simplify the common cases of optional parameter.

Link to Proposal #13

**Comment by jimbobmcgee** [comment]
This may only matter in relation to documentation, but in your examples you include the statements:

    Foo( Optional arg4 As Integer ) ' --> Foo( Optional arg4 As Integer = Nothing )
    Foo( Optional arg5 As Integer = 0 ) ' --> Foo( Optional arg5 As Integer = 0 )

If I understand correctly, those statements are identical, since `Dim i as Integer = Nothing` is functionally equivalent to C#'s `default` keyword, not its `null` keyword (appreciating that this isn't a C# discussion).

As such, technically, the example might be:

    Foo( Optional arg4 As Integer? ) ' --> Foo( Optional arg4 As Integer? = Nothing )
    Foo( Optional arg5 As Integer )  ' --> Foo( Optional arg5 As Integer = 0 )

For the avoidance of doubt, I'm not suggesting that this adversely affects the *proposal*, only its *documentation*.  

As someone who occasionally switches between teams that work in C# *and* VB, I know that the *expectation* of `Nothing` being functionaly equivalent to `null` catches people out more than you might hope.

**Comment by AdamSpeight2008** [comment]
@jimbobmcgee
In vb I think of `Nothing` as equivalent to `Default(Of T)`. 
```
Class Type => Nothing (aka null reference)
Structure Type => Nothing ( bit pattern of structure is all zeros)
Nullable Type =>  .HasValue := False, .Value:= 0 ( InvalidOperation_NoValue exception if called)
```
The documentation is correct in terms of those examples. They are demonstrating what the feature actually get lowered to. I didn't put in an example with a nullable type, as can derived from the structure type example.

**Comment by jimbobmcgee** [comment]
@AdamSpeight2008
I agree that `Nothing` is equivalent to `default(T)`, but I've seen many people forget it in the wild.  To that end (and only that end), I was pointing out where the examples you had given might confuse the issue.

In terms of what Roslyn/the compiler *does* -- lowering, etc. -- I am not qualified to comment.  If your documentation is intended only to cover that, e.g. for internal use, you'll get no argument from me!

But for any public documentation of the feature, I would *not* imply a difference between `Nothing` and `0` as in your `arg4`/`arg5` examples above; and I *would* introduce an example that covers a Nullable type.

**Comment by AdamSpeight2008** [comment]
@jimbobmcgee 
I agree that public documentation should be more complete. My proposals are more on technical side, and assume a little knowledge of how it may be implemented in the compiler.

**Comment by MarcusMM2** [comment]
I would like to have the keyword "optional" itself optional:
optional i as integer = 0       ->      i as integer = 0
the "= 0" indicates that it must be optional

**Comment by AdamSpeight2008** [comment]
@MarcusMM2 Could you make that a separate proposal as the referred to proposal has already be approved by VB / Roslyn Team.
